### ISSUE TITLE
consensus: proposer-based timestamps

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#2840]
Current mechanism for BFT time includes a timestamp in every precommit, and just takes the median. 

While simple and live despite arbitrary clock drift, it has some problems:

- can't be used for [signature aggregation](https://github.com/tendermint/tendermint/issues/1319) (since each vote has a unique timestamp component)
- gives +1/3 of the validator set complete control over the timestamp (see https://github.com/tendermint/tendermint/issues/2653)

An alternative, originally proposed in https://github.com/tendermint/tendermint/pull/1615, is to let the proposer use their own local timestamp, and for validators to only vote for the block if the timestamp is within some range of their own clock. This requires all nodes to keep their clocks reasonably closely synced, and was originally rejected/postponed for this reason. But in order to solve the above two issues, we should reconsider this solution. We can also include an option to disable it, so folks that don't want to worry about synced clocks won't have to (but of course it means they might not have accurate timestamps in blocks).

We should write a new ADR for this and do some more analysis before fully moving forward with it.

Note there is also a follow up issue from the last BFT-time PR: https://github.com/tendermint/tendermint/issues/2309

**Comment by mdyring** [comment]
Slack chatter:

**bucky**
One important thing to note about #2840 (part of v0.33 milestone) is that it will introduce a clock synchronization assumption into Tendermint - ie. validators will be expected to run NTP or equivalent so that their internal clocks are roughly synchronized, else there will be a liveness failure. This is quite different from Tendermint now, and is the reason we didn’t move forward with this last summer when we introduced BFT time. Currently we just take the median of the timestamps, so there is no clock sync assumption, but the new approach will make it so proposed blocks are only valid if the timestamp is within some delta around each validators local timestamp. If you have concerns with this, please comment in the issue! (edited)

**Ethan Frey**
As long as delta is reasonably large (like 2 minutes, not 2 seconds), this seems like a reasonable assumption to me.

Tendermint halts with network failures, so there are plenty of assumptions about connectivity already

**mdyring**
Tried to read up on this, perhaps below was already addressed:

What happens once validators (out of malice or clock drift), starts proposing blocks that are right at the upper boundary (future) of true time+delta? Lets call this timestamp F.

Next proposer in turn might have a more accurate view of time, and won't be able to propose a block with timestamp before F (as time must move forward), resulting in missed block.

Seems this could also be exploited to make it hard/impossible for a next block proposer to choose a time (by deliberately proposing with a time in the future). (edited)

**bucky**
yep this is the kind of stuff we need to be super careful about and why we need more detailed spec write up first

**mdyring**
Somehow my intuition is that delta should be < block time for this to work, otherwise a validator can create shenanigans for next proposer. I would also expect NTP sync'd servers to be within a few hundred msec worst case.

**Comment by ebuchman** [comment]
Take heed of whats in #2309. It may be made irrelevant by the work here though.

**Comment by ebuchman** [comment]
> gives +1/3 of the validator set complete control over the timestamp (see #2653)

This allows +1/3 to attack lite clients by forwarding the time past their unbonding period. This can be mitigated by making the unbonding period a hybrid of time and blocks. Related: https://github.com/tendermint/tendermint/issues/2565.

**Comment by melekes** [comment]
`A previous PR (#1615) implemented a proposer based time that would change Tendermint's liveness guarantees.` (https://github.com/tendermint/tendermint/issues/2013#issue-342691163)

**Comment by melekes** [comment]
> Next proposer in turn might have a more accurate view of time, and won't be able to propose a block with timestamp before F (as time must move forward), resulting in missed block.

this is unless you allow block time to drift backwards

**Comment by tessr** [comment]
@milosevic, @josef-widder, @ebuchman, @melekes, @marbar3778, @cwgoes and I just had a call about this, to try to figure out next steps and also to figure out which releases this work might need to target. Most urgently: Is this needed for light client security (read: IBC launch)? 

It's worth noting that discussions about "time" in Tendermint are really often talking about three distinct strategies: proposer-based non-BFT time, which was the previous implementation; median BFT time, which is the current implementation; and proposer-based BFT time, which is where we'd like to be eventually. 

We would like to eventually adopt proposer-based BFT time for two reasons:

1. It is necessary for BLS/signature aggregation 
2. It increases light client security; in particular, it protects against an attack where someone with >1/3 validators can change the unbonding period (and then do something bad and run away unslashed) 

The downside of adopting proposer-based BFT time is that it imposes stricter requirements on validators around clock synchronization and may also change Tendermint's guarantees around termination. Note that both proposer-based non-BFT time and median BFT time produce timestamps as almost a byproduct of the consensus process, but that the timestamps are not _part_ of the consensus process. That would change in a proposer-based BFT time world. 

In other words, there's some more spec/verification work to be done before we can implement proposer-based BFT time. 

This is okay, and we've decided to not try to include this work in 0.34, because the two motivations for proposer-based BFT time are not urgent and/or can be mitigated:

1. The BLS/signature aggregation issue is non-urgent.
2. Although light client security is necessary for IBC launch, we have a mitigation strategy which has already been merged to the Cosmos SDK (more on that below). Additionally, the >1/3 unbonding period attack may not be the most urgent concern, as someone with >1/3 validator power could already suppress evidence handling. 

For these reasons, we've decided to exclude proposer-based BFT time from the upcoming 0.34 release, and will work with Informal to spec/verify this approach at some point in the future. Moving forward, there are a number of product-y/protocol-y questions to investigate:

* Who are the users who would be impacted by different guarantees (and higher requirements around clock synchronization)? 
* Would some applications prefer this level of synchronization requirement (non-time oracles, for example)?
* Are there possibilities of different "flavors" of Tendermint that make different guarantees and tradeoffs around time/termination/liveness? 

In the meantime, we can mitigate the light client security risks by making "time" a hybrid of clock time and block height in the Cosmos SDK. If the unbonding period gets moved too much, the "blocks" part of "time" will kick in. (One caveat: the light client can't rely only on height to know if the timestamp is correct.) [This change has already been implemented in the Cosmos SDK](https://github.com/cosmos/cosmos-sdk/pull/5952).

#### Prior Discussions

* BFT Time (1): https://github.com/tendermint/tendermint/issues/1146 
* BFT Time (2): https://github.com/tendermint/tendermint/issues/2013 
* Follow up on BFT Time https://github.com/tendermint/tendermint/issues/2309 
* Proposer-based timestamps https://github.com/tendermint/tendermint/issues/2840 
* Current BFT Time spec https://github.com/tendermint/spec/blob/master/spec/consensus/bft-time.md

**Comment by piersy** [comment]
Hi @ebuchman, @melekes, @tessr 

I'm wondering how this proposal is going to impact the validity check defined in the tendermint consensus algorithm as described in the tendermint [whitepaper](https://arxiv.org/pdf/1807.04938.pdf.).

I had previously considered the function call `valid(v)` to be objective, I.E every network participant with the same state would receive the same results when executing `valid(v)`.

This suggestion to move to proposer based timestamps looks like it would change the nature of calls to `valid(v)` to be subjective based on a participant's current view of time.

Looking at the upon condition defined at line 49 of the pseudocode (the condition to agree a value) defined in the [whitepaper](https://arxiv.org/pdf/1807.04938.pdf.) it seems that nodes whose clocks are far out of sync could get stuck if they do not consider the value valid. If 2f+1 participants move to the next height, nodes that did not consider the value valid will be locked out of consensus unable to ever rejoin.

One potential way I can see to solve this would be to remove the validity check at this point. If 2f+1 nodes have precommitted for a value that appears invalid, then either the network has exceeded the level of tolerable failures/byzantine nodes, in which case it's already broken, or those nodes may actually be right.

That does feel a bit unsafe to me though, since it introduces the possibility of valid nodes being coerced into agreeing to totally invalid values, so maybe a better way would be to define 2 validity functions.

`objectivelyValid(v)` and `subjectivelyValid(v)`.

`objectivelyValid(v)` would check only the elements that all correct participants are guaranteed to agree on.
`subjectivelyValid(v)` for now would just check the time.

Then we could say that nodes will never vote for an objectively invalid value but they may agree to what they perceive as a subjectively invalid value if they see 2f+1 or more precommits for it.

**Comment by cmwaters** [comment]
The team has decided to go forward with implementing proposer based timestamps sometime this quarter. The folks at informal have already helped write up an [extensive spec](https://github.com/tendermint/spec/tree/master/spec/consensus/proposer-based-timestamp). I'm going to pin this issue to track the progress. As most of the details have been sufficiently covered, we'll look to first write up a small ADR before getting started on the implementation

**Comment by josef-widder** [comment]
I am happy to review (or contribute to) the ADR for the Informal side. @cmwaters, please don't hesitate to ping me whenever you think I can be of any help!  
During July, I might not be able to respond quickly, but from August, I should be more responsive ;-)

**Comment by cmwaters** [comment]
@williambanfield can this issue be finally closed ? :)
